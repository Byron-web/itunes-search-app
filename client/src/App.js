import React, { useState } from "react";
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import NavBar from "./components/navbar";
import Search from "./components/search";
import Favourites from "./components/favourites";

function App() {
  const [favourites, setFavourites] = useState(
    JSON.parse(sessionStorage.getItem("favourites")) || []
  );

  const handleAddToFavourites = (item) => {
    setFavourites([...favourites, item]);
  };

  const handleRemoveFromFavourites = (item) => {
    const updatedFavourites = favourites.filter((fav) => fav.id !== item.id);
    setFavourites(updatedFavourites);
  };

  const saveFavouritesToStorage = (favourites) => {
    sessionStorage.setItem("favourites", JSON.stringify(favourites));
  };

  // Listen to changes in favourites state and save to session storage
  React.useEffect(() => {
    saveFavouritesToStorage(favourites);
  }, [favourites]);

  console.log(sessionStorage);
  return (
    <Router>
      <div className="App">
        <NavBar />
        <Routes>
          <Route
            path="/"
            element={<Search handleAddToFavourites={handleAddToFavourites} />}
          />
          <Route
            path="/favourites"
            element={
              <Favourites
                favourites={favourites}
                handleRemoveFromFavourites={handleRemoveFromFavourites}
              />
            }
          />
        </Routes>
      </div>
    </Router>
  );
}

export default App;
