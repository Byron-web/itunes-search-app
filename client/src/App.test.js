import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import App from "./App";
import Search from "./components/search";

// Test that the App component renders correctly
test("renders App component", () => {
  const { asFragment } = render(<App />);
  // Render the App component and take a snapshot of the rendered output
  expect(asFragment()).toMatchSnapshot();
});

// Test that search results are updated when a search is performed
test("it should update the search results on search", async () => {
  // Mock search results
  const mockResults = [
    {
      trackId: 1,
      artworkUrl100: "https://example.com/image1.png",
      trackName: "Track 1",
      artistName: "Artist 1",
      longDescription: "This is a long description",
      shortDescription: "This is a short description",
    },
    {
      trackId: 2,
      artworkUrl100: "https://example.com/image2.png",
      trackName: "Track 2",
      artistName: "Artist 2",
      longDescription: "This is a long description",
      shortDescription: "This is a short description",
    },
  ];

  const mockFetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({ results: mockResults }),
    })
  );
  global.fetch = mockFetch;

  // Render the Search component
  render(<Search />);

  // Simulate a search by changing the value of the search input and clicking the search button
  const searchInput = screen.getByPlaceholderText("Search");
  fireEvent.change(searchInput, { target: { value: "query" } });

  const searchButton = screen.getByText("Search");
  fireEvent.click(searchButton);

  // Verify that fetch is called with the correct URL
  expect(mockFetch).toHaveBeenCalledTimes(1);
  expect(mockFetch).toHaveBeenCalledWith(
    "https://itunes.apple.com/search?term=query&entity=song"
  );

  // Verify that the search results are displayed correctly
  const cardTitles = await screen.findAllByTestId("card-title");
  expect(cardTitles).toHaveLength(2);
  expect(cardTitles[0]).toHaveTextContent("Track 1");
  expect(cardTitles[1]).toHaveTextContent("Track 2");
});
