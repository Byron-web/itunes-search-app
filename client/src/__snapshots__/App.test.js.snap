// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App component 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="navbar navbar-expand-lg navbar-light bg-light"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand"
          href="#home"
        >
          iTunes Search
        </a>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              class="nav-link"
              href="/"
            >
              Search
            </a>
            <a
              class="nav-link"
              href="/favourites"
            >
              Favourites
            </a>
          </div>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <form
        class=""
      >
        <div>
          <input
            class="form-control"
            id="search"
            placeholder="Search"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            class="form-label"
            for="type"
          >
            Search Type:
          </label>
          <select
            class="form-control"
            id="type"
          >
            <option
              value="all"
            >
              All
            </option>
            <option
              value="movie"
            >
              Movies
            </option>
            <option
              value="podcast"
            >
              Podcasts
            </option>
            <option
              value="song"
            >
              Music
            </option>
            <option
              value="audiobook"
            >
              Audiobooks
            </option>
            <option
              value="shortFilm"
            >
              Short Films
            </option>
            <option
              value="tvEpisode"
            >
              TV Shows
            </option>
            <option
              value="software"
            >
              Software
            </option>
            <option
              value="ebook"
            >
              eBooks
            </option>
          </select>
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
      <hr />
      <div
        class="row"
      />
    </div>
  </div>
</DocumentFragment>
`;
